utils/src/mesh.rs

```diff
@@ -5,109 +5,136 @@ use kaboodle::{errors::KaboodleError, Kaboodle};
-pub trait KaboodlePeer {
-    fn new(address: SocketAddr, identity: Vec<u8>) -> Self;
-    fn identity(&self) -> Vec<u8>;
-    fn name(&self) -> &str;
-    fn roles(&self) -> Vec<String>;
-}
-
 /// A little wrapper around kaboodle so we can hide the machinery of encoding and decoding.
 /// the identity payload.
 #[async_trait]
 pub trait KaboodleMesh {
     type A;
 
     /// Create a new entry for a Kaboodle peer network and add ourselves to the mesh.
-    async fn new(
-        name: &str,
-        roles: Vec<String>,
-        port: u16,
-        interface: Option<Interface>,
-    ) -> Result<Box<Self>, KaboodleError>;
+    async fn start(&mut self) -> Result<(), KaboodleError>;
     /// Get a list of peers. It's the implementer's responsibility to decide if this is fresh or cached somehow.
     async fn peers(&self) -> Vec<Self::A>;
-    /// Given a specific role, look for a peer that advertises the role.
-    async fn find_role(&self, role: &str) -> Result<Self::A, KaboodleError>;
 }
 
+/// This type encodes the responsibilities of the resources we are meshing together.
+pub trait KaboodlePeer {
+    /// Create a new peer structure from the node identity payload plus an address.
+    fn from_identity(address: SocketAddr, encoded: Vec<u8>) -> Self;
+    /// Create an identity payload from whatever internal information matters to your implementation.
+    fn identity(&self) -> Vec<u8>;
+    /// Get the address of this node.
+    fn address(&self) -> Option<SocketAddr>;
+}
+
+// End of tiny wrapper around Kaboodle.
+
+// An envelope that holds a version number. A little bit of future-proofing
+// to allow agents with higher version numbers to decode payloads from older agents.
+// Possibly over-thinking it.
 #[derive(Debug, Clone, Decode, Encode)]
 struct VersionEnvelope {
     version: u8,
     rest: Vec<u8>
 }
 
+#[derive(Debug, Clone)]
+pub struct PeerMetadata {
+    address: Option<SocketAddr>,
+    inner: MetadataInner,
+}
+
+// The ddta we need to encode our identity as a serval peer. Done with an additional
+// type to get the derive. There'll be another way to do this, I'm sure.
 #[derive(Debug, Clone, Decode, Encode)]
-pub struct ServalPeer {
+struct MetadataInner {
     name: String,
-    roles: Vec<String>,
+    roles: Vec<String>
+}
+
+impl PeerMetadata {
+    pub fn new(name: String, roles: Vec<String>, address: Option<SocketAddr>) -> Self {
+        let inner = MetadataInner { name, roles };
+        Self {
+            address, inner
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.inner.name
+    }
+
+    pub fn roles(&self) -> Vec<String> {
+        self.inner.roles.clone()
+    }
 }
 
-impl KaboodlePeer for ServalPeer {
-    fn new(address: SocketAddr, encoded: Vec<u8>) -> Self {
-        // TODO: this is actually fallible
+impl KaboodlePeer for PeerMetadata {
+    fn from_identity(address: SocketAddr, encoded: Vec<u8>) -> Self {
+        // TODO: this is actually fallible; when might it fail?
```

Comment:
This data is coming in over unreliable transport, so it's possible for a bit to get flipped and lead to a corrupted encoded payload.